File: library/src/main/java/com/thefinestartist/finestwebview/listeners/BroadCastManager.java
package com.thefinestartist.finestwebview.listeners;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.support.annotation.NonNull;
import android.support.v4.content.LocalBroadcastManager;

import java.util.List;

/**
 * Created by TheFinestArtist on 1/26/16.
 */
public class BroadCastManager {

    public static final String WEBVIEW_EVENT = "WEBVIEW_EVENT";
    public static final String EXTRA_KEY = "EXTRA_KEY";
    public static final String EXTRA_TYPE = "EXTRA_TYPE";
    public static final String EXTRA_URL = "EXTRA_URL";
    public static final String EXTRA_TITLE = "EXTRA_TITLE";
    public static final String EXTRA_PROGESS = "EXTRA_PROGESS";
    public static final String EXTRA_PRECOMPOSED = "EXTRA_PRECOMPOSED";

    protected int key;
    protected List<WebViewListener> listeners;
    protected LocalBroadcastManager manager;
    protected BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override public void onReceive(Context context, Intent intent) {
            if (context == null || intent == null) return;
            int key = intent.getIntExtra(EXTRA_KEY, Integer.MIN_VALUE);
            if (BroadCastManager.this.key == key) handleIntent(intent);
        }
    };

    public BroadCastManager(Context context, int key, @NonNull List<WebViewListener> listeners) {
        this.key = key;
        this.listeners = listeners;
        manager = LocalBroadcastManager.getInstance(context);
        manager.registerReceiver(receiver, new IntentFilter(WEBVIEW_EVENT));
    }

    public enum Type {
        PROGRESS_CHANGED, RECEIVED_TITLE, RECEIVED_TOUCH_ICON_URL, PAGE_STARTED, PAGE_FINISHED, LOAD_RESOURCE, PAGE_COMMIT_VISIBLE, UNREGISTER
    }

    private void handleIntent(@NonNull Intent intent) {
        Type type = (Type) intent.getSerializableExtra(EXTRA_TYPE);
        switch (type) {
            case PROGRESS_CHANGED:
                onProgressChanged(intent);
                break;
            case RECEIVED_TITLE:
                onReceivedTitle(intent);
                break;
            case RECEIVED_TOUCH_ICON_URL:
                onReceivedTouchIconUrl(intent);
                break;
            case PAGE_STARTED:
                onPageStarted(intent);
                break;
            case PAGE_FINISHED:
                onPageFinished(intent);
                break;
            case LOAD_RESOURCE:
                onLoadResource(intent);
                break;
            case PAGE_COMMIT_VISIBLE:
                onPageCommitVisible(intent);
                break;
            case UNREGISTER:
                unregister();
                break;
        }
    }

    // Base Static Methods
    private static Intent getBaseIntent(int key, Type type) {
        return new Intent(BroadCastManager.WEBVIEW_EVENT)
                .putExtra(EXTRA_KEY, key)
                .putExtra(EXTRA_TYPE, type);
    }

    private static void sendBroadCast(Context context, Intent intent) {
        LocalBroadcastManager.getInstance(context).sendBroadcast(intent);
    }

    // Handle Each Event Type
    public static void onProgressChanged(Context context, int key, int progress) {
        Intent intent = getBaseIntent(key, Type.PROGRESS_CHANGED).putExtra(EXTRA_PROGESS, progress);
        sendBroadCast(context, intent);
    }

    public void onProgressChanged(Intent intent) {
        for (WebViewListener listener : listeners)
            listener.onProgressChanged(intent.getIntExtra(EXTRA_PROGESS, 0));
    }

    public static void onReceivedTitle(Context context, int key, String title) {
        Intent intent = getBaseIntent(key, Type.RECEIVED_TITLE).putExtra(EXTRA_TITLE, title);
        sendBroadCast(context, intent);
    }

    public void onReceivedTitle(Intent intent) {
        for (WebViewListener listener : listeners)
            listener.onReceivedTitle(intent.getStringExtra(EXTRA_TITLE));
    }

    public static void onReceivedTouchIconUrl(Context context, int key, String url, boolean precomposed) {
        Intent intent = getBaseIntent(key, Type.RECEIVED_TOUCH_ICON_URL).putExtra(EXTRA_URL, url).putExtra(EXTRA_PRECOMPOSED, precomposed);
        sendBroadCast(context, intent);
    }

    public void onReceivedTouchIconUrl(Intent intent) {
        for (WebViewListener listener : listeners)
            listener.onReceivedTouchIconUrl(intent.getStringExtra(EXTRA_URL), intent.getBooleanExtra(EXTRA_PRECOMPOSED, false));
    }

    public static void onPageStarted(Context context, int key, String url) {
        Intent intent = getBaseIntent(key, Type.PAGE_STARTED).putExtra(EXTRA_URL, url);
        sendBroadCast(context, intent);
    }

    public void onPageStarted(Intent intent) {
        for (WebViewListener listener : listeners)
            listener.onPageStarted(intent.getStringExtra(EXTRA_URL));
    }

    public static void onPageFinished(Context context, int key, String url) {
        Intent intent = getBaseIntent(key, Type.PAGE_FINISHED).putExtra(EXTRA_URL, url);
        sendBroadCast(context, intent);
    }

    public void onPageFinished(Intent intent) {
        for (WebViewListener listener : listeners)
            listener.onPageFinished(intent.getStringExtra(EXTRA_URL));
    }

    public static void onLoadResource(Context context, int key, String url) {
        Intent intent = getBaseIntent(key, Type.LOAD_RESOURCE).putExtra(EXTRA_URL, url);
        sendBroadCast(context, intent);
    }

    public void onLoadResource(Intent intent) {
        for (WebViewListener listener : listeners)
            listener.onLoadResource(intent.getStringExtra(EXTRA_URL));
    }

    public static void onPageCommitVisible(Context context, int key, String url) {
        Intent intent = getBaseIntent(key, Type.PAGE_COMMIT_VISIBLE).putExtra(EXTRA_URL, url);
        sendBroadCast(context, intent);
    }

    public void onPageCommitVisible(Intent intent) {
        for (WebViewListener listener : listeners)
            listener.onPageCommitVisible(intent.getStringExtra(EXTRA_URL));
    }

    public static void unregister(Context context, int key) {
        Intent intent = getBaseIntent(key, Type.UNREGISTER);
        sendBroadCast(context, intent);
    }

    private void unregister() {
        if (manager != null && receiver != null)
            manager.unregisterReceiver(receiver);
    }
}

File: library/src/main/java/com/thefinestartist/finestwebview/listeners/BroadCastManager.kt
package com.thefinestartist.finestwebview.listeners

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import androidx.localbroadcastmanager.content.LocalBroadcastManager

/**
 * Created by TheFinestArtist on 1/26/16.
 */
class BroadCastManager(
  context: Context,
  private val key: Int,
  private val listeners: List<WebViewListener>
) {
  private val manager: LocalBroadcastManager = LocalBroadcastManager.getInstance(context)
  private val receiver: BroadcastReceiver = object : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
      val key = intent.getIntExtra(EXTRA_KEY, Int.MIN_VALUE)
      if (this@BroadCastManager.key == key) {
        handleIntent(intent)
      }
    }
  }

  init {
    manager.registerReceiver(receiver, IntentFilter(WEBVIEW_EVENT))
  }

  private fun handleIntent(intent: Intent) {
    when (intent.getSerializableExtra(EXTRA_TYPE)) {
      Type.PROGRESS_CHANGED -> onProgressChanged(intent)
      Type.RECEIVED_TITLE -> onReceivedTitle(intent)
      Type.RECEIVED_TOUCH_ICON_URL -> onReceivedTouchIconUrl(intent)
      Type.PAGE_STARTED -> onPageStarted(intent)
      Type.PAGE_FINISHED -> onPageFinished(intent)
      Type.LOAD_RESOURCE -> onLoadResource(intent)
      Type.PAGE_COMMIT_VISIBLE -> onPageCommitVisible(intent)
      Type.DOWNLOADED_START -> onDownloadStart(intent)
      Type.UNREGISTER -> unregister()
    }
  }

  private fun onProgressChanged(intent: Intent) {
    for (listener in listeners) {
      listener.onProgressChanged(intent.getIntExtra(EXTRA_PROGRESS, 0))
    }
  }

  private fun onReceivedTitle(intent: Intent) {
    for (listener in listeners) {
      listener.onReceivedTitle(intent.getStringExtra(EXTRA_TITLE))
    }
  }

  private fun onReceivedTouchIconUrl(intent: Intent) {
    for (listener in listeners) {
      listener.onReceivedTouchIconUrl(
        intent.getStringExtra(EXTRA_URL),
        intent.getBooleanExtra(EXTRA_PRECOMPOSED, false)
      )
    }
  }

  private fun onPageStarted(intent: Intent) {
    for (listener in listeners) {
      listener.onPageStarted(intent.getStringExtra(EXTRA_URL))
    }
  }

  private fun onPageFinished(intent: Intent) {
    for (listener in listeners) {
      listener.onPageFinished(intent.getStringExtra(EXTRA_URL))
    }
  }

  private fun onLoadResource(intent: Intent) {
    for (listener in listeners) {
      listener.onLoadResource(intent.getStringExtra(EXTRA_URL))
    }
  }

  private fun onPageCommitVisible(intent: Intent) {
    for (listener in listeners) {
      listener.onPageCommitVisible(intent.getStringExtra(EXTRA_URL))
    }
  }

  private fun onDownloadStart(intent: Intent) {
    for (listener in listeners) {
      listener.onDownloadStart(
        intent.getStringExtra(EXTRA_URL),
        intent.getStringExtra(EXTRA_USER_AGENT),
        intent.getStringExtra(EXTRA_CONTENT_DISPOSITION),
        intent.getStringExtra(EXTRA_MIME_TYPE),
        intent.getLongExtra(EXTRA_CONTENT_LENGTH, 0L)
      )
    }
  }

  private fun unregister() {
    manager.unregisterReceiver(receiver)
  }

  enum class Type {
    PROGRESS_CHANGED,
    RECEIVED_TITLE,
    RECEIVED_TOUCH_ICON_URL,
    PAGE_STARTED,
    PAGE_FINISHED,
    LOAD_RESOURCE,
    PAGE_COMMIT_VISIBLE,
    DOWNLOADED_START,
    UNREGISTER
  }

  companion object {
    const val WEBVIEW_EVENT = "WEBVIEW_EVENT"
    const val EXTRA_KEY = "EXTRA_KEY"
    const val EXTRA_TYPE = "EXTRA_TYPE"
    const val EXTRA_URL = "EXTRA_URL"
    const val EXTRA_TITLE = "EXTRA_TITLE"
    const val EXTRA_PROGRESS = "EXTRA_PROGRESS"
    const val EXTRA_PRECOMPOSED = "EXTRA_PRECOMPOSED"
    const val EXTRA_USER_AGENT = "EXTRA_USER_AGENT"
    const val EXTRA_CONTENT_DISPOSITION = "EXTRA_CONTENT_DISPOSITION"
    const val EXTRA_MIME_TYPE = "EXTRA_MIME_TYPE"
    const val EXTRA_CONTENT_LENGTH = "EXTRA_CONTENT_LENGTH"

    // Base Static Methods
    private fun getBaseIntent(key: Int, type: Type): Intent {
      return Intent(WEBVIEW_EVENT).putExtra(EXTRA_KEY, key).putExtra(EXTRA_TYPE, type)
    }

    private fun sendBroadCast(context: Context, intent: Intent) {
      LocalBroadcastManager.getInstance(context).sendBroadcast(intent)
    }

    // Handle Each Event Type
    @JvmStatic
    fun onProgressChanged(context: Context, key: Int, progress: Int) {
      val intent = getBaseIntent(key, Type.PROGRESS_CHANGED).putExtra(EXTRA_PROGRESS, progress)
      sendBroadCast(context, intent)
    }

    @JvmStatic
    fun onReceivedTitle(context: Context, key: Int, title: String?) {
      val intent = getBaseIntent(key, Type.RECEIVED_TITLE).putExtra(EXTRA_TITLE, title)
      sendBroadCast(context, intent)
    }

    @JvmStatic
    fun onReceivedTouchIconUrl(context: Context, key: Int, url: String?, precomposed: Boolean) {
      val intent =
        getBaseIntent(key, Type.RECEIVED_TOUCH_ICON_URL)
          .putExtra(EXTRA_URL, url)
          .putExtra(EXTRA_PRECOMPOSED, precomposed)
      sendBroadCast(context, intent)
    }

    @JvmStatic
    fun onPageStarted(context: Context, key: Int, url: String?) {
      val intent = getBaseIntent(key, Type.PAGE_STARTED).putExtra(EXTRA_URL, url)
      sendBroadCast(context, intent)
    }

    @JvmStatic
    fun onPageFinished(context: Context, key: Int, url: String?) {
      val intent = getBaseIntent(key, Type.PAGE_FINISHED).putExtra(EXTRA_URL, url)
      sendBroadCast(context, intent)
    }

    @JvmStatic
    fun onLoadResource(context: Context, key: Int, url: String?) {
      val intent = getBaseIntent(key, Type.LOAD_RESOURCE).putExtra(EXTRA_URL, url)
      sendBroadCast(context, intent)
    }

    @JvmStatic
    fun onPageCommitVisible(context: Context, key: Int, url: String?) {
      val intent = getBaseIntent(key, Type.PAGE_COMMIT_VISIBLE).putExtra(EXTRA_URL, url)
      sendBroadCast(context, intent)
    }

    @JvmStatic
    fun onDownloadStart(
      context: Context,
      key: Int,
      url: String?,
      userAgent: String?,
      contentDisposition: String?,
      mimeType: String?,
      contentLength: Long
    ) {
      val intent =
        getBaseIntent(key, Type.DOWNLOADED_START)
          .putExtra(EXTRA_URL, url)
          .putExtra(EXTRA_USER_AGENT, userAgent)
          .putExtra(EXTRA_CONTENT_DISPOSITION, contentDisposition)
          .putExtra(EXTRA_MIME_TYPE, mimeType)
          .putExtra(EXTRA_CONTENT_LENGTH, contentLength)
      sendBroadCast(context, intent)
    }

    @JvmStatic
    fun unregister(context: Context, key: Int) {
      val intent = getBaseIntent(key, Type.UNREGISTER)
      sendBroadCast(context, intent)
    }
  }
}